doc/Design.md

```diff
@@ -143,3 +143,91 @@ dereferencing. Unlike shared buffers, which can only be a buffer type in a
 capsule, granted memory can be defined as any type. Therefore, processes cannot
 access this memory since doing so might violate type-safety.
 
+## Some in-kernel design principles
+
+### Role of HILs
+
+Generally, the Tock kernel is structured into three layers:
+
+1. Chip-specific drivers: these typically live in a crate in the
+   `chips` subdirectory, or an equivalent crate in an different repo
+   (e.g. the Titan port is out of tree but it’s `h1b` create is the
+   equivalent here). These drivers have implements that are specific
+   to the hardware particular to a certain microcontroller. Ideally,
+   their implementation is fairly simple, and they merely adhere to a
+   common interface (a HIL). That’s not always the case, but that’s
+   the ideal.
+
+2. Chip-agnostic, portable, peripheral drivers and subsystems. These
+   typically live in the `capsules` crate. These includes things like
+   the virtual alarms and virtual I2C stack, as well as drivers for
+   hardware peripherals not on the chip itself (e.g. sensors, radios,
+   etc). These drivers typically rely on the chip-specific drivers
+   through the HILs.
+
+3. System call drivers, also typically found in the `capsules`
+   crate. These are the drivers that implement a particular part of
+   the system call interfaces, and are often even more abstracted from
+   the hardware than (2) - for example, the temperature sensor system
+   call driver can use any temperature sensor, including several
+   implemented as portable peripheral drivers. We don’t have many
+   examples of this, but the system call interface is another point of
+   standardization that can be implemented in various ways. So it’s
```

Comment:
Arguably the lack of this in Tock hints at having multiple implementations of the same `Driver` is unnecessary. Instead, only one thing implements the actual syscall interface, and then that can depend on any hardware.

Reply:
We will almost certainly have an example of this for a Console over USB once the USB for NRF52 fully lands... and in general, I think this is actually an important design point for Tock, regardless of whether the “blessed” version in this repo has examples of it.

Reply:
As an aside, this decomposition suggests to me that we should break out syscalls/ from capsules/. Or have it be a submodule. I.e., have this conceptual structure mirrored in the directory structure.